### PR TITLE
query-frontend: Allow for multiple topics in readConsistencyRoundTripper

### PR DESCRIPTION
**Pull request by leizor**
#### What this PR does

Adjust `readConsistencyRoundTripper` to take a map of header keys to offsets readers, allowing it to support the tracking of the produced offsets of multiple topics.

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Review by pracucci — APPROVED** [review]
Nice work, LGTM!

Please rebase with `main`. I've done some small changes to reader metrics, but shouldn't be an issue to merge with your changes.

File: pkg/storage/ingest/reader.go [hunk 1]
```diff
@@ -121,12 +121,16 @@ func newPartitionReader(kafkaCfg KafkaConfig, partitionID int32, instanceID stri
...
 	}
 
-	r.metrics = newReaderMetrics(partitionID, reg, r)
+	r.metrics = newReaderMetrics(partitionID, reg, r, kafkaCfg.Topic)
 
 	r.Service = services.NewBasicService(r.start, r.run, r.stop)
 	return r, nil
 }
 
+func (r *PartitionReader) Topic() string {
```
[the remark this answers is not in the document]
Comment (leizor) [reply pkg/storage/ingest/reader.go]: You're correct - accidentally left this from an abandoned attempt at adding the topic label to the read consistency metrics. Removed in d3c9d1d79389491270529b33242d5750116ccba0!

File: pkg/mimir/modules.go [hunk 2]
```diff
@@ -726,8 +727,13 @@ func (t *Mimir) initQueryFrontendTopicOffsetsReader() (services.Service, error)
 		return t.IngesterPartitionRingWatcher.PartitionRing().PartitionIDs(), nil
 	}
 
-	t.QueryFrontendTopicOffsetsReader = ingest.NewTopicOffsetsReader(kafkaClient, t.Cfg.IngestStorage.KafkaConfig.Topic, getPartitionIDs, t.Cfg.IngestStorage.KafkaConfig.LastProducedOffsetPollInterval, t.Registerer, util_log.Logger)
-	return t.QueryFrontendTopicOffsetsReader, nil
+	ingestTopicOffsetsReader := ingest.NewTopicOffsetsReader(kafkaClient, t.Cfg.IngestStorage.KafkaConfig.Topic, getPartitionIDs, t.Cfg.IngestStorage.KafkaConfig.LastProducedOffsetPollInterval, t.Registerer, util_log.Logger)
+
+	t.QueryFrontendTopicOffsetsReaders = map[string]*ingest.TopicOffsetsReader{
```
[the remark this answers is not in the document]
Comment (leizor) [reply pkg/mimir/modules.go]: I was planning to add the new aggregations topic offsets reader module as dependent on this `QueryFrontendTopicOffsetsReaders` module, but it does seem safer to do the check you suggested - added in 68bc274ef11cef6e9daba04ee55da2de9f0e501f.

**Comment by leizor** [comment]
Thanks for the review @pracucci!

**Pull request merged**